### PR TITLE
Remove redundant `.float()` in Attention

### PR DESCRIPTION
**Pull request by eaccgod**
The `.float()` in the Attention layer is not necessary because `F.softmax` internally casts everything to float32, then operates on it, then casts back down to whatever precision you're operating in. Proof:

```
# python3
Python 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>>
root@da4e9fb38af4:/llama# python3
Python 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> import torch.nn.functional as F
>>> orig = torch.randn(100000, device="cuda", dtype=torch.bfloat16)
>>> softmaxed = F.softmax(orig, dim=-1)
>>> softmaxed_float = F.softmax(orig.float(), dim=-1).type_as(orig)
>>> (softmaxed_float - softmaxed).abs().sum()
tensor(0., device='cuda:0', dtype=torch.bfloat16)
```

Removing the `.float()` meant I could get a slightly higher batch size without OOMing.

### CODE REVIEW

File: llama/model.py [hunk 1]
```diff
@@ -177,7 +177,7 @@ def forward(
         scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(self.head_dim)
         if mask is not None:
             scores = scores + mask  # (bs, n_local_heads, seqlen, cache_len + seqlen)
-        scores = F.softmax(scores.float(), dim=-1).type_as(xq)
+        scores = F.softmax(scores.type_as(xq), dim=-1).type_as(xq)
```
Comment (javier-m) [inline llama/model.py:180]: as `xq` and `scores` are both `float16`, and as `F.softmax` returns a `float16` when inputted a `float16`, why not simplifying by

```
scores = F.softmax(scores, dim=-1)
```
